lit_parrot/speed_monitor.py

```diff
@@ -0,0 +1,272 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+# Adapted for standalone use
+from collections import deque
+from contextlib import nullcontext
+from typing import Deque, Optional
+
+import torch
+from lightning.fabric.loggers import Logger
+from torch.utils.flop_counter import FlopCounterMode
+
+from lit_parrot import Parrot
+
+GPU_AVAILABLE_FLOPS = {
+    # source: https://resources.nvidia.com/en-us-tensor-core/nvidia-tensor-core-gpu-datasheet
+    # nvidia publishes spec sheet with a 2x sparsity factor
+    "h100-sxm": {
+        "64-true": 67e12,
+        "32-true": 67e12,
+        "16-true": 1.979e15 / 2,
+        "16-mixed": 1.979e15 / 2,
+        "bf16-true": 1.979e15 / 2,
+        "bf16-mixed": 1.979e15 / 2,
+        "8-true": 3.958e15 / 2,
+        "8-mixed": 3.958e15 / 2,
+    },
+    "h100-pcie": {
+        "64-true": 51e12,
+        "32-true": 51e12,
+        "16-true": 1.513e15 / 2,
+        "16-mixed": 1.513e15 / 2,
+        "bf16-true": 1.513e15 / 2,
+        "bf16-mixed": 1.513e15 / 2,
+        "8-true": 3.026e15 / 2,
+        "8-mixed": 3.026e15 / 2,
+    },
+    # source: https://www.nvidia.com/content/dam/en-zz/Solutions/Data-Center/a100/pdf/nvidia-a100-datasheet-us-nvidia-1758950-r4-web.pdf
+    # sxm and pcie have same flop counts
+    "a100": {
+        "64-true": 19.5e12,
+        "32-true": 19.5e12,
+        "16-true": 312e12,
+        "16-mixed": 312e12,
+        "bf16-true": 312e12,
+        "bf16-mixed": 312e12,
+    },
+    # source: https://images.nvidia.com/content/technologies/volta/pdf/volta-v100-datasheet-update-us-1165301-r5.pdf
+    "v100-sxm": {"64-true": 7.8e12, "32-true": 15.7e12, "16-true": 125e12, "16-mixed": 125e12},
+    "v100-pcie": {"64-true": 7e12, "32-true": 14e12, "16-true": 112e12, "16-mixed": 112e12},
+    "v100s-pcie": {"64-true": 8.2e12, "32-true": 16.4e12, "16-true": 130e12, "16-mixed": 130e12},
+    # source: https://www.nvidia.com/content/dam/en-zz/Solutions/Data-Center/tesla-t4/t4-tensor-core-datasheet-951643.pdf
+    # sxm and pcie have same flop counts
+    "t4": {"32-true": 8.1e12, "16-true": 65e12, "16-mixed": 65e12, "8-true": 130e12, "int4": 260e12},
+    # https://www.nvidia.com/content/dam/en-zz/Solutions/design-visualization/quadro-product-literature/quadro-rtx-5000-data-sheet-us-nvidia-704120-r4-web.pdf
+    "quadro rtx 5000": {"32-true": 11.2e12, "16-true": 89.2e12, "16-mixed": 89.2e12},
+}
+
+
+def get_gpu_flops_available(precision: str):
+    # Return 0 if no CUDA device (e.g., when running with CPU only)
+    if not torch.cuda.is_available():
+        return 0
```

Comment:
We could make this function get the device type so that it can work for XLA devices too. @gkroiz Do you know where we can get the FLOPs for the different TPU archs?

Reply:
Hi, https://cloud.google.com/tpu/docs/system-architecture-tpu-vm seems to have some numbers for v2, v3, and v4, but only for bfloat16. Let me see if I can find a document that covers FLOPs for other levels of precision